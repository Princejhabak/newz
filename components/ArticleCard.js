import moment from "moment";
import Link from 'next/link'

// Component to show a single news article with the below details
const ArticleCard = ({article}) => {

    const {source, published_at, title, description, author, category, image, url} = article;

    return(
        <div className="row m-4 article-card">
            {
                image ?
                <>
                    <div className="col-9 py-2 article-card-left">
                        <div className="article-card-header">
                            <h5 className="article-card-header-source">{source}</h5>
                            {/* <h6 className="text-warning">{moment(published_at).utcOffset("+00:00").format('MMMM Do, YYYY')}</h6> */}
                            <h6 className="text-warning">{moment(published_at).utcOffset("+00:00").format('MMMM Do, YYYY HH:mm')}</h6>
                            {/* <h6 className="text-warning">{published_at}</h6> */}
                        </div>
                        <h5 className="article-card-heading text-dark">
                            <a href={url} target="blank">
                                {title}
                            </a>
                        </h5>
                        <small className="article-card-body">{description}</small>
                        <div className="article-card-footer">
                            <small className="text-warning"><b>Author: {author ? author : 'N/A'}</b></small>
                            <small className=""><b>Category: {category}</b></small>
                        </div>
                    </div>
                    <div className="col-3 article-card-right">
                            <img src={image} className="img-fluid rounded-start article-card-img" alt="image" /> 
                    </div>
                </>
                :
                <>
                    <div className="col-12 py-2 article-card-left">
                        <div className="article-card-header">
                            <h5 className="article-card-header-source">{source}</h5>
                            <h6 className="text-warning">{moment(published_at).utcOffset("+00:00").format('MMMM Do, YYYY HH:mm')}</h6>
                        </div>
                        <h5 className="article-card-heading text-dark">
                            <a href={url} target="blank">
                                {title}
                            </a>
                        </h5>
                        <small className="article-card-body">{description}</small>
                        <div className="article-card-footer">
                            <small className="text-warning"><b>Author: {author ? author : 'N/A'}</b></small>
                            <small className=""><b>Category: {category}</b></small>
                        </div>
                    </div>
                </>
            }
        </div>
    );
}

export default ArticleCard;
