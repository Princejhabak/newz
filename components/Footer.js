import { Container, Row, Col } from 'react-bootstrap';

// Footer component for the app
const Footer = () => {
  return (
    <footer>
      <Container>
        <Row>
          <Col className='text-center py-3'>Copyright &copy; Newz</Col>
        </Row>
      </Container>
    </footer>
  )
}

export default Footer;

