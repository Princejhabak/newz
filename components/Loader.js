import { Spinner } from 'react-bootstrap'

// Loading component for the app - shows at every async task
const Loader = () => {
  return (
    <Spinner
      animation='border'
      role='status'
      style={{
        width: '100px',
        height: '100px',
        margin: 'auto',
        display: 'block',
      }}
    >
      <span className='sr-only'>Loading...</span>
    </Spinner>
  )
}

export default Loader;