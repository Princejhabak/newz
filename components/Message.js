import React from "react";
import { Alert } from 'react-bootstrap'

// Custom message component for the app to display and handle server errors
const Message = ({ variant, children }) => {
  return <Alert data-testid="alert" variant={variant}>{children}</Alert>
}

Message.defaultProps = {
  variant: 'info',
}

export default Message