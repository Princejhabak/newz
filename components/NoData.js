
// Component to show a no data image and other details when the aricle list is empty
const NoData = () => {
    return(
        <div className="no-data-container">
            <h3>Oops!</h3>
            <img src="no_data.png" alt="No Data" className="no-data-img mt-3 mb-5"/>
            <h3>No Articles Found</h3>
        </div>
    )
}

export default NoData;