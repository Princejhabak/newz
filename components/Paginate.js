import Pagination from 'react-bootstrap/Pagination';
import { useRouter } from 'next/router';

// Custom pagination component for the app
// Props - current page, pagination obj received from server
const Paginate = ({page, pagination}) => {

    const router = useRouter();

    if(!pagination){
        return <></>;
    }

    // Calculate total no of available pages
    const totalPages = pagination.total/pagination.limit;

    const pages = [];
    const min = Math.min(5, totalPages-page+1);
   
    // Create an array of max 5 items starting from the current page
    for(let i=0; i<min; i++){
        pages.push(page+i);
    }

    // Helper function to handle pagination routing
    const handleClick = (pg) => {
        let url = router.asPath;
        let arr = url.split("&");

        if(arr[arr.length-1].includes("page=")){
            arr = arr.slice(0,arr.length-1);
            url = arr.join("&")
        }
        if(url === "/news"){
            url = url + '?'
        }
        url = url + `&page=${pg}`;
        router.push(url);
    }

    return (
        <Pagination className="mt-5 paginate flex-row justify-content-center px-auto" size="lg">
            <Pagination.First disabled={page===1} onClick={() => handleClick(1)}/>
            <Pagination.Prev disabled={page===1} onClick={() => handleClick(page-1)}/>
            {
                pages.map(item => <Pagination.Item active={item===page} key={item} onClick={() => handleClick(item)}>{item}</Pagination.Item>)
            }
            <Pagination.Next disabled={page===totalPages} onClick={() => handleClick(page+1)}/>
            <Pagination.Last disabled={page===totalPages} onClick={() => handleClick(totalPages)}/>
        </Pagination>
    );
}

export default Paginate;