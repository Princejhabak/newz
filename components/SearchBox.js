import { useState } from 'react';
import { Form, Button } from 'react-bootstrap';
import { useRouter } from 'next/router';

// Custom search component for the app's header/navbar
const SearchBox = () => {

  const [keyword, setKeyword] = useState('');

  const router = useRouter();

//   const submitHandler = (e) => {
//     e.preventDefault();
//     router.push(`/news?search=${keyword}`);
//   }

  const handleClick = (e) => {
    e.preventDefault();
    router.push(`/news?search=${keyword}`);
  }

  return (
    <Form inline data-testid="nav-search-form">
    {/* <Form onSubmit={submitHandler} inline data-testid="nav-search-form"></Form> */}
      <Form.Control
        type='text'
        name='q'
        onChange={(e) => setKeyword(e.target.value)}
        placeholder='Search Articles...'
        className='mr-sm-2 ml-sm-5'
      />
      <SearchButton disabled={false} onClick={handleClick}>
        Search
      </SearchButton>
    </Form>
  )
}

export default SearchBox;

export const SearchButton = ({disabled, onClick, children}) => {
    return(
        <Button 
            type='submit' 
            variant='outline-warning' 
            className='navbar-search-btn'
            disabled={disabled}
            onClick={onClick}
            data-testid="nav-search-button"
        >
            {children}
        </Button>
    )
}