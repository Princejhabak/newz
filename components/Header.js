import { useDispatch, useSelector } from 'react-redux';
import { Navbar, Nav, Container, NavDropdown } from 'react-bootstrap';
import SearchBox from './SearchBox';
import { changeLanguage } from '../actions/articleActions';
import {LANGUAGES} from '../utils/metadata';
import { useRouter } from 'next/router';
import Link from 'next/link';

// Header/Navbar component for the app
const Header = () => {

  const router = useRouter();
  const dispatch = useDispatch();

  // Get language from centeral redux state
  const {language} = useSelector((state) => state.language);

  // Helper function to change the language from dropdown
  const handleLanguageChange = (eventKey, event) => {
    dispatch(changeLanguage(eventKey));
  }

  return (
      <header>
        <Navbar className="navbar" bg='dark' variant='dark' expand='lg' collapseOnSelect>
          <Container>
                <Navbar.Brand className='navbar-brand' data-testid="nav-brand">
                    <Link href="/">
                        Newz
                    </Link>
                </Navbar.Brand>
                
              <Navbar.Toggle aria-controls='basic-navbar-nav' />
              <Navbar.Collapse id='basic-navbar-nav'>
                  <SearchBox  />
                  <Nav className='ml-auto'>
                    <Nav.Link className="mr-3" onClick={() => router.push('/news')}>
                        Live News
                    </Nav.Link>
                    <Nav.Link className="mr-3" onClick={() => router.push('/news?&categories=sports')}>
                        Sports
                    </Nav.Link>
                    <Nav.Link className="mr-3" onClick={() => router.push('/news?&categories=health')}>
                        Health
                    </Nav.Link>
                    <NavDropdown title={language} id="collasible-nav-dropdown" onSelect={handleLanguageChange}>
                        {
                        LANGUAGES.map(language => <NavDropdown.Item key={language.key} eventKey={language.key}>{language.value}</NavDropdown.Item> )
                        }
                    </NavDropdown>
                  </Nav>
              </Navbar.Collapse>
          </Container>
        </Navbar>
      </header>
    )
}

export default Header;
