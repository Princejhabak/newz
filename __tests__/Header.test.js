/* eslint-disable */

import React from "react";
import {SearchButton} from "../components/SearchBox";
import { fireEvent, render, screen, cleanup } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";

afterEach(cleanup);

describe("Search Box", () => {

    it("button exists", () => {
        render(<SearchButton>Search</SearchButton> )

        const searchButton = screen.getByTestId("nav-search-button");
        expect(searchButton).toBeInTheDocument();
    });

    it("id disabled when needed", () => {
        const handleClick = jest.fn();

        render(<SearchButton onClick={handleClick} disabled={true}>Search</SearchButton>);

        const searchButton = screen.getByTestId("nav-search-button");
        fireEvent.click(searchButton);
        expect(handleClick).toHaveBeenCalledTimes(0);
    });

    it("runs a function when clicked", () => {
        const handleClick = jest.fn();

        render(<SearchButton onClick={handleClick}>Search</SearchButton>);

        const searchButton = screen.getByTestId("nav-search-button");
        fireEvent.click(searchButton);
        expect(handleClick).toHaveBeenCalledTimes(1);
    });

    it("matches snapshot", () => {
        const tree = renderer.create(<SearchButton>Search</SearchButton>).toJSON();
        expect(tree).toMatchSnapshot();
    });

});