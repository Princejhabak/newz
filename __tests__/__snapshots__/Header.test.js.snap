// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Box matches snapshot 1`] = `
<button
  className="navbar-search-btn btn btn-outline-warning"
  data-testid="nav-search-button"
  disabled={false}
  type="submit"
>
  Search
</button>
`;
